xmr-btc/src/lib.rs

```diff
@@ -50,3 +50,151 @@ pub mod bitcoin;
 pub mod bob;
 pub mod monero;
 pub mod transport;
+
+use ecdsa_fun::{adaptor::Adaptor, nonce::Deterministic};
+use genawaiter::sync::{Gen, GenBoxed};
+use sha2::Sha256;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub enum Action {
+    LockBitcoin(bitcoin::TxLock),
+    SendBitcoinRedeemEncsig(bitcoin::EncryptedSignature),
+    CreateMoneroWalletForOutput {
+        spend_key: monero::PrivateKey,
+        view_key: monero::PrivateViewKey,
+    },
+    RefundBitcoin {
+        tx_cancel: bitcoin::Transaction,
+        tx_refund: bitcoin::Transaction,
+    },
+}
+
+// TODO: This could be moved to the monero module
+pub trait ReceiveTransferProof {
+    fn receive_transfer_proof(&self) -> monero::TransferProof;
+}
+
+/// Perform the on-chain protocol to swap monero and bitcoin as Bob.
+///
+/// This is called post handshake, after all the keys, addresses and most of the
+/// signatures have been exchanged.
+pub fn action_generator_bob<N, M, B>(
+    network: &'static N,
+    monero_ledger: &'static M,
+    bitcoin_ledger: &'static B,
+    // TODO: Replace this with a new, slimmer struct?
+    bob::State2 {
+        A,
+        b,
+        s_b,
+        S_a_monero,
+        S_a_bitcoin,
+        v,
+        xmr,
+        refund_timelock,
+        redeem_address,
+        refund_address,
+        tx_lock,
+        tx_cancel_sig_a,
+        tx_refund_encsig,
+        ..
+    }: bob::State2,
+) -> GenBoxed<Action, (), ()>
+where
+    N: ReceiveTransferProof + Send + Sync,
+    M: monero::CheckTransfer + Send + Sync,
+    B: bitcoin::WatchForRawTransaction + Send + Sync,
+{
+    Gen::new_boxed(|co| async move {
+        let swap_result: Result<(), ()> = {
+            co.yield_(Action::LockBitcoin(tx_lock.clone())).await;
+
+            // the source of this could be the database, this layer doesn't care
+            let transfer_proof = network.receive_transfer_proof();
+
+            let S_b_monero = monero::PublicKey::from_private_key(&monero::PrivateKey::from_scalar(
+                s_b.into_ed25519(),
+            ));
+            let S = S_a_monero + S_b_monero;
+
+            // TODO: We should require a specific number of confirmations on the lock
+            // transaction
+            monero_ledger
+                .check_transfer(S, v.public(), transfer_proof, xmr)
+                .await
+                .expect("TODO: implementor of this trait must make it infallible by retrying");
+
+            let tx_redeem = bitcoin::TxRedeem::new(&tx_lock, &redeem_address);
+            let tx_redeem_encsig = b.encsign(S_a_bitcoin.clone(), tx_redeem.digest());
+
+            co.yield_(Action::SendBitcoinRedeemEncsig(tx_redeem_encsig.clone()))
```

Comment:
If this is network communication, you might want to yield the message directly, just so that the "driver" on the outside is a stupid as possible and more logic is contained within the protocol execution.

Reply:
> you might want to yield the message directly

Do you mean the message type, as opposed to its contents (in this case an encrypted signature)? If so, I agree.

Reply:
Yes the actual message type, just so that this component contains all the logic of what should happen when (which is the essence of a protocol) and the outside component can be dumb and only really does IO without having to convert between types.